RecoHGCal/TICL/plugins/GNNInterpretationAlgo.cc

```diff
@@ -0,0 +1,592 @@
+                                       const edm::ESHandle<MagneticField> bfieldH,
+                                       const edm::ESHandle<Propagator> propH) {
+  hgcons_ = hgcons;
+  rhtools_ = rhtools;
+  bfield_ = bfieldH;
+  propagator_ = propH;
+
+  buildLayers();
+}
+// Geometry construction
+void GNNInterpretationAlgo::buildLayers() {
+  // Build propagation disks at:
+  //  - HGCal front face
+  //  - CE-E CE-H interface
+
+  const float z_front = hgcons_->waferZ(1, true);
+  const auto r_front = hgcons_->rangeR(z_front, true);
+
+  const float z_interface = rhtools_.getPositionLayer(rhtools_.lastLayerEE()).z();
+  const auto r_interface = hgcons_->rangeR(z_interface, true);
+
+  for (int side = 0; side < 2; ++side) {
+    const float sign = (side == 0 ? -1.f : 1.f);
+
+    firstDisk_[side] = std::make_unique<GeomDet>(
+        Disk::build(Disk::PositionType(0, 0, sign * z_front),
+                    Disk::RotationType(),
+                    SimpleDiskBounds(r_front.first, r_front.second, sign * z_front - 0.5f, sign * z_front + 0.5f))
+            .get());
+
+    interfaceDisk_[side] = std::make_unique<GeomDet>(
+        Disk::build(Disk::PositionType(0, 0, sign * z_interface),
+                    Disk::RotationType(),
+                    SimpleDiskBounds(
+                        r_interface.first, r_interface.second, sign * z_interface - 0.5f, sign * z_interface + 0.5f))
+            .get());
+  }
+}
+
+// Trackster propagation
+Vector GNNInterpretationAlgo::propagateTrackster(const Trackster& t,
+                                                 unsigned idx,
+                                                 float zVal,
+                                                 std::array<TICLLayerTile, 2>& tracksterTiles) {
+  // needs only the positive Z co-ordinate of the surface to propagate to
+  // the correct sign is calculated inside according to the barycenter of trackster
+
+  // Propagation direction
+  const Vector& barycenter = t.barycenter();
+
+  // NOTE:
+  // barycenter as direction for tracksters w/ poor PCA
+  // propagation still done to get the cartesian coords
+  // which are anyway converted to eta, phi in linking
+  // -> can be simplified later
+
+  //FP: disable PCA propagation for the moment and fallback to barycenter position
+  // if (t.eigenvalues()[0] / t.eigenvalues()[1] < 20)
+
+  Vector direction = barycenter.unit();
+
+  // Ensure correct Z-side propagation
+  zVal *= (barycenter.Z() > 0.f ? 1.f : -1.f);
+
+  const float scale = (zVal - barycenter.Z()) / direction.Z();
+  const Vector propPoint(scale * direction.X() + barycenter.X(), scale * direction.Y() + barycenter.Y(), zVal);
+
+  // Fill spatial tiles for fast lookup
+  const bool isPositiveZ = (propPoint.Eta() > 0.f);
+  tracksterTiles[isPositiveZ].fill(propPoint.Eta(), propPoint.Phi(), idx);
+
+  return propPoint;
+}
+
+std::pair<float, float> GNNInterpretationAlgo::CalculateTrackstersError(const Trackster& trackster) {
```

Comment:
I think CMSSW naming rules require the method name to start with lower case letter